drivers/sd/bcm2711/brcme88c/README.md

```diff
@@ -0,0 +1,29 @@
+# Broadcom SD Host (BRCME88C) for Raspberry Pi 4
+
+Driver for the eMMC2 host controller (BRCME88C) for Raspberry Pi 4.
+
+This is a miniport - it depends on the in-box sdport.sys to handle the
+SD/SDIO/eMMC protocol and WDM interfaces.
+
+Implemented using the
+[Microsoft SDHC driver sample](https://github.com/microsoft/Windows-driver-samples/tree/master/sd/miniport/sdhc)
+as a reference.
+
+## NOTES
+
+- If both bcm2836sdhc and this driver are loaded and one is unloaded, the
+  system will eventually crash. Appears to be a bug in sdport.sys's interrupt
+  handling (the Arasan and eMMC2 controllers share an interrupt vector).
+  - Possible workaround: Don't use both bcm2836sdhc and this driver.
+  - Possible workaround: Don't unload drivers.
+
+## TODO
+
+- Regulator voltage cannot be controlled while in crashdump mode. Possible fix
+  would be to have mailbox support via ACPI method instead of via RPIQ driver.
```

Comment:
There's also a driver load order issue with RPIQ. If this driver loads before RPIQ, it won't be able to run at 50MB mode.

Reply:
I'll ask the UEFI devs if it's feasible to have a _DEP for RPIQ in EMMC2's ACPI descriptor. This might affect other OSes that don't have ACPI bindings for the mailbox driver or don't require it for voltage switching.

Reply:
@idigdoug From the RPIQ driver's INF, RPIQ looks to be marked as a boot start driver 
https://github.com/raspberrypi/windows-drivers/blob/31c2c748e6358711cd72467b7ad52da08691e54b/drivers/mailbox/bcm2836/RPIQ.inf#L55
and this EMMC2 driver is currently marked as a standard driver start so IIRC RPIQ should always be started before the EMMC2 driver. Have you observed differently in your bringup?

That being said, I can imagine this EMMC2 driver becoming a boot critical driver eventually, assuming we want to boot from it in the future 😃. I'm not exactly sure how to enforce boot critical driver dependencies but it should be doable (famous last words).

Note: It's been a while since I've dug into Windows PnP rules so my memory might be hazy...

Reply:
I've been focused on making the driver code work, so I haven't done much testing of various scenarios. I haven't played around with booting from the SD card yet. However, the way BootCritical is supposed to work nowadays is that the driver is supposed to be promoted to BootCritical dynamically. In the case of this driver, it will be promoted to BootCritical if the system is booting from an SD card.

I don't want the driver to fail to load if RPIQ is unavailable, so I don't want to declare that this driver has a direct dependency on RPIQ. (Otherwise this driver could never work in SafeOS for upgrade scenarios.) It might be possible to do something with load order groups, though that's not my area of expertise.

My understanding of using an ACPI method for controlling mailbox is that adding the ACPI method won't force anybody to use ACPI for mailbox, so just adding ACPI methods for mailbox wouldn't directly affect Linux. However, each OS should pick exactly one way to access mailbox -- you shouldn't have one component access mailbox via ACPI and another component access mailbox via RPIQ.sys. So it should be ok to have ACPI methods for mailbox even if Linux doesn't use them, but if we do have ACPI methods for mailbox, all of the Windows drivers should only use ACPI methods.

Reply:
https://docs.microsoft.com/en-us/windows-hardware/drivers/install/specifying-driver-load-order

Note that I've set BootFlags = CM_SERVICE_SD_DISK_BOOT_LOAD, so this driver becomes boot-start when booting from SD.

Reply:
"Boot Bus Extender" appears to come before "System Bus Extender", so maybe it will just work.

Reply:
As long as the driver is able to work without RPIQ (but at slower speeds as you've mentioned), it's okay for now. The proper fix is to replace the code calling RPIQ with an ACPI method. sdbus for example can call an Intel-specific _DSM for voltage switching. We could use that (so that sdbus can benefit from it too), or think of another way (a PowerResource perhaps?).

Until that will be sorted out, my idea was to mark RPIQ as a dependency for EMMC2 in ACPI (this would enforce the correct load order no matter what), but the driver won't load if RPIQ is not installed.

Reply:
The driver should definitely be configured so it can load even when RPIQ is not available. I definitely don't want a dependency. I think LoadOrderGroup will fix this as much as it can be fixed.

Reply:
Just tried doing a clean install with the driver. Windows almost reaches the "Getting ready" phase, but the screen goes black instantly. I'll try again tomorrow with a debugger attached.

Reply:
Not sure what was wrong with my previous installation, but I've been able to successfully complete the setup this time.

RPIQ always gets loaded before SDHostBRCME88C:
```
ModLoad: fffff803`800b0000 fffff803`800c7000   vdrvroot.sys
ModLoad: fffff803`800d0000 fffff803`800dd000   rpiq.sys
ModLoad: fffff803`800e0000 fffff803`8010e000   pdc.sys 
ModLoad: fffff803`80110000 fffff803`80127000   CEA.sys 
ModLoad: fffff803`80130000 fffff803`8015e000   partmgr.sys
ModLoad: fffff803`80160000 fffff803`8021b000   spaceport.sys
ModLoad: fffff803`80220000 fffff803`8023a000   volmgr.sys
ModLoad: fffff803`80240000 fffff803`802a4000   volmgrx.sys
ModLoad: fffff803`802b0000 fffff803`802fb000   sdbus.sys
ModLoad: fffff803`80300000 fffff803`8030e000   SDHostBRCME88C.sys
ModLoad: fffff803`80310000 fffff803`80330000   sdport.sys
```